trace/backend_collector_test.go

```diff
@@ -452,3 +499,114 @@ func TestOnlyObfuscateNonErrorResponses(t *testing.T) {
+				"Content-Type": {"application/json"},
+				"postman_sid":  {"XX__X_X_X_X"},
+			},
+			Cookies: []*http.Cookie{
+				{
+					Name:  "Random-Cookie",
+					Value: "Random-Cookie-Value",
+				},
+			},
+			Body: memview.New([]byte(`{"homes": ["error", "happened", "here"], "pmakInResponseBody": "PMAK-6717875c69335700017b1c46"}`)),
+		},
+	}
+
+	col := NewBackendCollector(fakeSvc, fakeLrn, mockClient, optionals.None[int](), NewPacketCounter(), true, nil)
+	assert.NoError(t, col.Process(errReq))
+	assert.NoError(t, col.Process(errResp))
+	assert.NoError(t, col.Close())
+
+	expectedWitnesses := []*pb.Witness{
+		{
+			Method: &pb.Method{
+				Id: &pb.MethodID{
+					ApiType: pb.ApiType_HTTP_REST,
+				},
+				Args: map[string]*pb.Data{
+					"4F1vWo8G_-Q=": newTestHeaderSpec(dataFromPrimitive(spec_util.NewPrimitiveString("")), "x-access-token", 0),
+					"KC2RO-pCNJA=": newTestHeaderSpec(dataFromPrimitive(spec_util.NewPrimitiveString("Normal-Value")), "Normal-Header", 0),
+					"xwb2G1yYVVc=": newTestHeaderSpec(dataFromPrimitive(spec_util.NewPrimitiveString("")), "pmak_in_header", 0),
+					"9NijbeQiJAg=": newTestQueryParamSpec(dataFromPrimitive(spec_util.NewPrimitiveString("")), "sso_jwt_key", 0),
+					"b5t-IaNo7Ug=": newTestQueryParamSpec(dataFromPrimitive(spec_util.NewPrimitiveString("")), "pmak_in_query", 0),
+					"k5p4y9tXMAk=": newTestAuthSpec(dataFromPrimitive(spec_util.NewPrimitiveString("")), 0),
+					"bxitt4RTL5k=": newTestBodySpecFromStruct(0, pb.HTTPBody_JSON, "application/json", map[string]*pb.Data{
+						"name":       dataFromPrimitive(spec_util.NewPrimitiveString("error")),
+						"number":     dataFromPrimitive(spec_util.NewPrimitiveInt64(202410081550)),
+						"pmakInBody": dataFromPrimitive(spec_util.NewPrimitiveString("")),
+					}),
+				},
+				Responses: map[string]*pb.Data{
+					"hAjVb_ouhwQ=": newTestCookieSpec(dataFromPrimitive(spec_util.NewPrimitiveString("")), "Random-Cookie", 404),
+					"rZob7SB3qd0=": newTestHeaderSpec(dataFromPrimitive(spec_util.NewPrimitiveString("")), "postman_sid", 404),
+					"cBn6EHKhiAA=": newTestBodySpecFromStruct(404, pb.HTTPBody_JSON, "application/json", map[string]*pb.Data{
+						"homes": dataFromList(
+							dataFromPrimitive(spec_util.NewPrimitiveString("error")),
+							dataFromPrimitive(spec_util.NewPrimitiveString("happened")),
+							dataFromPrimitive(spec_util.NewPrimitiveString("here")),
+						),
+						"pmakInResponseBody": dataFromPrimitive(spec_util.NewPrimitiveString("")),
+					}),
+				},
+				Meta: &pb.MethodMeta{
+					Meta: &pb.MethodMeta_Http{
+						Http: &pb.HTTPMethodMeta{
+							Method:       "POST",
+							PathTemplate: "/v1/doggos",
+							Host:         "example.com",
+							Obfuscation:  pb.HTTPMethodMeta_NONE,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for i := range expectedWitnesses {
+		expected := proto.MarshalTextString(expectedWitnesses[i])
+		actual := proto.MarshalTextString(rec.witnesses[i])
+		assert.Equal(t, expected, actual)
```

Comment:
I believe assert.Equal will work on the original protobufs -- look at some other unit tests. Marshaling to string form should not be necessary, and I think assert.Equal will attempt to give a meaningful diff without it.

Reply:
Tried, it didn't work
```
expected: &api_spec.Witness{state:impl.MessageState{NoUnkeyedLiterals:pragma.NoUnkeyedLiterals{}, DoNotCompare:pragma.DoNotCompare{}, DoNotCopy:pragma.DoNotCopy{}, atomicMessageInfo:(*impl.MessageInfo)(nil)}, sizeCache:0, unknownFields:[]uint8(nil), Method:(*api_spec.Method)(0x140003b7e00)}

actual: &api_spec.Witness{state:impl.MessageState{NoUnkeyedLiterals:pragma.NoUnkeyedLiterals{}, DoNotCompare:pragma.DoNotCompare{}, DoNotCopy:pragma.DoNotCopy{}, atomicMessageInfo:(*impl.MessageInfo)(0x1400015e2c0)}, sizeCache:0, unknownFields:[]uint8(nil), Method:(*api_spec.Method)(0x140003b7270)}
```
`atomicMessageInfo` is different

Reply:
OK, sorry.  The example I was thinking of was from lib/learn:

```
var (
        cmpWitnessOptions = []cmp.Option{
                cmp.Comparer(proto.Equal),
                cmpopts.SortSlices(witnessLess),
                cmpopts.EquateEmpty(),
        }
)

....

                if diff := cmp.Diff(expected, candidates, cmpWitnessOptions...); diff != "" {
                        t.Errorf("[%s] found diff:\n %v", c.name, diff)
                }

```